Sources/MessagingInApp/State/InAppMessageMiddleware.swift

```diff
@@ -0,0 +1,247 @@
+        }
+
+        // Process message queue to check if there is a message that matches the new route
+        dispatch(.processMessageQueue(messages: Array(state.messagesInQueue)))
+    }
+}
+
+func messageStateMiddleware(logger: CioInternalCommon.Logger) -> InAppMessageMiddleware {
+    middleware { _, getState, next, action in
+        // Continue to next middleware if action is not loadMessage
+        guard case .loadMessage(let message, let position) = action else {
+            return next(action)
+        }
+
+        let state = getState()
+        // If there is a message currently displayed, block loading new message
+        guard !state.currentMessageState.isDisplayed else {
+            let currentMessage = state.currentMessageState.message?.describeForLogs ?? "nil"
+            return next(.reportError(message: "[InApp] Blocked loading message: \(message.describeForLogs) because another message is currently displayed or cancelled: \(currentMessage)"))
+        }
+
+        logger.debug("[InApp] Showing message: \(message.describeForLogs) with position: \(String(describing: position))")
+        // Show message on main thread to avoid unexpected crashes
+        DispatchQueue.main.async {
+            let messageManager = MessageManager(siteId: state.siteId, message: message)
+            messageManager.showMessage(position: position ?? .center)
+        }
+
+        return next(action)
+    }
+}
+
+private func logMessageView(logger: CioInternalCommon.Logger, state: InAppMessageState, message: Message) {
+    LogManager(siteId: state.siteId, dataCenter: state.dataCenter).logView(
+        message: message, userToken: state.userId
+    ) { response in
+        if case .failure(let error) = response {
+            logger.error("[InApp] Failed to log message view: \(error) for message: \(message.describeForLogs)")
+        }
+    }
+}
+
+func messageMetricsMiddleware(logger: CioInternalCommon.Logger) -> InAppMessageMiddleware {
+    middleware { _, getState, next, action in
+        let state = getState()
+        switch action {
+        case .displayMessage(let message):
+            let properties = message.gistProperties
+            // Log message view only if message is not persistent
+            if properties.persistent != true {
+                logger.debug("[InApp] Message shown, logging view for message: \(message.describeForLogs)")
+                logMessageView(logger: logger, state: state, message: message)
+            } else {
+                logger.debug("[InApp] Persistent message shown, not logging view for message: \(message.describeForLogs)")
+            }
+
+        case .dismissMessage(let message, let shouldLog, let viaCloseAction):
+            guard shouldLog else { return }
+
+            let properties = message.gistProperties
+            // Log message close only if message was dismissed via close action
+            if viaCloseAction {
+                if properties.persistent == true {
```

Comment:
same as above, we can use `message.gistProperties` directly?

Reply:
Updated